# Assignment2

#This my second BINF6210 assignment

#Focusing on exploring the resolution of COI gene and CytB gene in reconstructing phylogenetic tree of fox species specifically!!!

#Introduction :

  Prior to the common use in genetic sequencing to classify species, species identification was primarily based on morphological, physiological differences, and habitat preferences. It was initially assumed that species with greater evolutionary relatedness show higher similarities in these characteristics. However, with the indepth research in science, the idea of convergent evolution has gained recognition. The idea states that species can exhibit similar morphological characteristics independently regardless of a common ancestor due to exposure to similar selective pressure. Hence, genetic sequencing can circumvent the idea of convergent evolution providing a relatively unbiased species classification method, since it utilize the idea that genetic blueprints are unique to individuals. The two most commonly used genes in animal phylogenetic studies are cytochrome oxidase I (COI) gene and cytochrome B (CytB) gene, which are both from mitochondrial DNA (mtDNA). Mitochondrial DNA has been known to be inherited maternally in mammals (Taanman 1999), this provides advantage in phylogenetic sequencing allowing for easy tracking of ancestors. Also, mtDNA encounters higher mutation rate compared to nuclear DNA because the mitochondria generates energy source for the cell while producing by product of reactive oxygen species (ROS). These ROS can oxidize the mtDNA in close proximity causing base mutation to occur more easily (Brown et al, 1971). However, in terms of phylogenetic sequencing the combination of high mutation and maternal inheritance allows mtDNA to serve as an effective marker for tracking genetic variability in species level.


  However, in current phylogenetic studies, there is no form of standardised approach for classifying species. This often leads to inconsistent evolutionary interpretation at species level due to the use of different genetic markers. Thus this study aims to contribute insights to support a standardised approach to phylogenetic reconstruction to determine which mitochondrial gene serves as the most effective marker at species level. The study compares cytochrome oxidase I and cytochrome B genes at species level within the Canidae family, with particular focus on fox-like canids. The Canidae family encompasses 35 species, such as wolves, coyotes, and foxes. They inhabit diverse ecosystems across all continents (Pallida & Hilton, 2014). For this analysis, only the genera of Otocyon, Vulpes, Lycalopex, Cerdocyon, and Urocyon are considered.




#Discussion/ Conclusion :

  The results suggest that the cytochrome B gene offers better resolution for species-level taxonomic classification, as it forms clearer clusters in dendrograms, indicating genetic similarities among species. In comparison, while the cytochrome oxidase I gene can similarly distinguish species relationships, it lacks the clustering nodes (genus level) that cytochrome B provides. Thus, the result aligns with the other research results in suggesting cytochrome B gene can provide better resolution for separating species (Tobe et al., 2010; Hsieh et al., 2001).


  Some limitations were encountered during the analysis of these nucleotide sequences. The data obtained from the BOLD system exhibits high variability with a noticeable skewed
direction towards very short and very long sequence length. This may have affected the quality of sequence alignment leading to less accurate comparison among the species. Consequently, this could have led to challenges in distance matrix clustering, where the nucleotide sequence differences were not distinctive enough to introduce genus nodes. Thus causing lower species resolution seen from the result compared to CytB gene. Additionally, both NCBI and BOLD databases show biases in the data samples they provide. It was noticeable that certain species are overrepresented in these databases leading to a larger number of samples for a specific taxa. Also, these taxa tend to have relatively longer and higher quality nucleotide sequences, leading to biases and eventually affect the quality of the result. For future projects, I would obtain and combine samples from different databases to help reduce these biases and enhance the quality of sequence alignment.
